lib/internal/modules/cjs/loader.js

```diff
@@ -1277,9 +1288,37 @@ function loadESMFromCJS(mod, filename) {
   const source = getMaybeCachedSource(mod, filename);
   const cascadedLoader = require('internal/modules/esm/loader').getOrInitializeCascadedLoader();
   const isMain = mod[kIsMainSymbol];
-  // TODO(joyeecheung): we may want to invent optional special handling for default exports here.
-  // For now, it's good enough to be identical to what `import()` returns.
-  mod.exports = cascadedLoader.importSyncForRequire(filename, source, isMain);
+  const namespace = cascadedLoader.importSyncForRequire(filename, source, isMain);
+  // Tooling in the ecosystem have been using the __esModule property to recognize
+  // transpiled ESM in consuming code. For example, a 'log' package written in ESM:
+  //
+  // export default function log(val) { console.log(val); }
+  //
+  // Can be transpiled as:
+  //
+  // exports.__esModule = true;
+  // exports.default = function log(val) { console.log(val); }
+  //
+  // The consuming code may be written like this in ESM:
+  //
+  // import log from 'log'
+  //
+  // Which gets transpiled to:
+  //
+  // const _mod = require('log');
+  // const log = _mod.__esModule ? _mod.default : _mod;
+  //
+  // So to allow transpiled consuming code to recognize require()'d real ESM
+  // as ESM and pick up the default exports, we add a __esModule property by
+  // composing a new object with the namespace as prototype and an __esModule
+  // property. This is done like this because namespace objects are not
+  // extensible and proxies or synthetic module facades are more expensive than
+  // prototype lookups.
+  if (namespace.__esModule) {
+    mod.exports = namespace;
```

Comment:
Maybe
```suggestion
    mod.exports = ObjectCreate(namespace);
```
so that the result of `require()` esm is consistent (e.g. always an object whose prototype is the namespace object).

The performance impact should be irrelevant, because in practice this branch will only be ever taken when importing an ESM file that does `export * from "./a-cjs-file"` (because no ES modules have an actual `__esModule` export since it's not compatible with bundlers and transpilers).